dm/master/openapi.go

```diff
@@ -133,17 +133,85 @@ func (s *Server) DMAPIDeleteSource(ctx echo.Context, sourceName string) error {
 
 // DMAPIStartRelay url is:(POST /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStartRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.StartRelayRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	needUpdate := false
+	// update relay related in source cfg
+	if req.RelayBinlogName != nil && sourceCfg.RelayBinLogName != *req.RelayBinlogName {
+		sourceCfg.RelayBinLogName = *req.RelayBinlogName
+		needUpdate = true
+	}
+	if req.RelayBinlogGtid != nil && sourceCfg.RelayBinlogGTID != *req.RelayBinlogGtid {
+		sourceCfg.RelayBinlogGTID = *req.RelayBinlogGtid
+		needUpdate = true
+	}
+	if req.RelayDir != nil && sourceCfg.RelayDir != *req.RelayDir {
+		sourceCfg.RelayDir = *req.RelayDir
+		needUpdate = true
+	}
+	if needUpdate {
+		// update current source relay config before start relay
+		if err := s.scheduler.UpdateSourceCfg(sourceCfg); err != nil {
+			return err
+		}
+	}
+	return s.scheduler.StartRelay(sourceName, req.WorkerNameList)
 }
 
 // DMAPIStopRelay url is:(DELETE /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStopRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.StopRelayRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	return s.scheduler.StopRelay(sourceName, req.WorkerNameList)
 }
 
 // DMAPIGetSourceStatus url is:(GET /api/v1/sources/{source-id}/status).
 func (s *Server) DMAPIGetSourceStatus(ctx echo.Context, sourceName string) error {
-	return nil
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	var resp openapi.GetSourceStatusResponse
+	worker := s.scheduler.GetWorkerBySource(sourceName)
+	// current this source not bound to any worker
+	if worker == nil {
+		resp.Data = append(resp.Data, openapi.SourceStatus{SourceName: sourceName})
+		resp.Total = len(resp.Data)
+		return ctx.JSON(http.StatusOK, resp)
+	}
+	// get status from worker
+	workerRespList := s.getStatusFromWorkers(ctx.Request().Context(), []string{sourceName}, "", true)
+	for _, workerStatus := range workerRespList {
+		if workerStatus == nil {
+			// this should not happen unless the rpc in the worker server has been modified
+			return terror.ErrOpenAPICommonError.New("worker's query-status response is nil")
+		}
+		if !workerStatus.Result {
+			return terror.ErrOpenAPICommonError.New(workerStatus.Msg)
```

Comment:
if one worker is broken, we can see status of rest workers

Reply:
an error_msg field is added to the source status, so that if the worker returns a message with an error in it, it will be populated into that field in https://github.com/pingcap/dm/pull/2083/commits/06827359de44c3a163995c7567820c4943b9efe2